dao/sale_dao.py

```diff
@@ -0,0 +1,44 @@
+from model import Sale
+from dao import BookDao
+from typing import List
+from datetime import datetime
+
+class SaleDao:
+    """Banco de Dados de Vendas
+        Estrutura do banco: 
+        (
+            livro, 
+            venda(titulo do livro, quantidade), 
+            momento_do_cadastro
+        )
+    """
+
+    _list_sales: List[Sale] = []
+    _book_dao = BookDao()
+
+    def save(self, books_sale: List[Sale]) -> None:
+        self.book_exists(books_sale)
+        self.sale_information()
+    
+    def book_exists(self, books_sale: List[Sale]) -> None:
+        for sale in books_sale:
+            books = self._book_dao.find_by_title(sale.title_book)
+            for book in books:
+                if sale.title_book == book.title:
+                    self._list_sales.append((book, sale, datetime.now()))
+        if not self._list_sales:
+            raise Exception('Livro não consta em nosso acervo')
```

Comment:
Funciona, mas todo laço dentro de laço tende a ser problemático, ainda mais com a complexidade de um if dentro dele. Confesso que não entendi esse if not no fim do método.

Reply:
Aqui ainda me parece que em todo lugar que precisar checar se o livro existe, você sempre vai fazer esse for in com o find by title junto.

Reply:
O método está fazendo duas tarefaz: checando a existência dos livros e adicionando eles na list sales, o que faz com que o nome book_exists não indique de fato o comportamento do método.

Reply:
O  find by title implementado no card anterior faz o seguinte: retornará uma lista com todos os livros que possui alguma parte daquele nome que passei(onde o mínimo de caractere a ser passado é 2). Então se eu faço:
find_by_title(python), ele irá me retornar, por exemplo:
- python fluente
- python para ds
- python avançado
então após ele me retornar essa lista, eu vejo se o título exato que a pessoa passou existe na lista que o find_by_title me retornou. 

Com o find_by_title do modo que está(retornando uma lista), a tendencia é que tenha laço dentro de laço, uma vez que eu também passo uma lista de vendas, então tenho que percorrer a lista que passei e como o find_by_title me retorna uma lista, tenho que percorrê-la para saber se há o livro exato que a pessoa passou. Se eu passasse um por um, não teria isso, eu diminuiria um for.

Reply:
Talvez o uso do find_by_title não te ajude a resolver o problema de uma forma interessante. Isso por que para cada livro que se encaixe no termo da busca deste caso, seu laço precisa fazer mais uma interação.